File: ios/Dagritme/Kern/Assistent.swift
// School, voetbal, judo en de bso sturen berichten; die wil je niet overtikken.
// Plak er een in en er komt een lijstje uit dat je met een tik overneemt. Weet
// de uitlezer iets niet — welk kind in 1-2B zit bijvoorbeeld — dan vraagt hij
// ernaar, en het antwoord blijft als kenmerk bij het kind staan.
//
// Het echte uitlezen doet Claude, op /api/lees in de Worker: daar staat de
// sleutel van de api, en die hoort niet in iets wat iedereen kan openmaken.
import Foundation

struct Lading {
    var tekst: String
    var vandaag: String
    var ronde: Int
    var kinderen: [Persoon]

    var lijf: [String: Any] {
        [
            "tekst": tekst,
            "vandaag": vandaag,
            "ronde": ronde,
            "kinderen": kinderen.map {
                ["id": $0.id, "naam": $0.naam, "kenmerken": $0.kenmerken] as [String: Any]
            },
        ]
    }
}

struct Vraag {
    var sleutel: String
    var vraag: String
    var opties: [String]
    var meerkeuze: Bool
}

struct Voorstel: Identifiable {
    let id = UUID()
    var ding: Ding
    var bron: String
}

// Zegt de uitlezer zelf wat er mis is, dan is dat de melding — er hoeft geen
// 'het lukte niet' omheen.
struct Uitleesfout: LocalizedError {
    var bericht: String
    var vanServer: Bool
    var errorDescription: String? { bericht }
}

func vraagAssistent(_ lading: Lading) async throws -> Json {
    guard let adres = Configuratie.assistent else {
        throw Uitleesfout(bericht: "Er is nog geen adres voor de uitlezer.", vanServer: false)
    }
    var verzoek = URLRequest(url: adres)
    verzoek.httpMethod = "POST"
    verzoek.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if !Configuratie.sleutel.isEmpty {
        verzoek.setValue(Configuratie.sleutel, forHTTPHeaderField: "X-Routine-Sleutel")
    }
    verzoek.httpBody = try JSONSerialization.data(withJSONObject: lading.lijf)
    // Uitlezen duurt langer dan een gewoon verzoek: Claude denkt even na.
    verzoek.timeoutInterval = 90

    let (gegevens, antwoord) = try await URLSession.shared.data(for: verzoek)
    let uit = Json.lees(gegevens)
    if let http = antwoord as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
        let reden = uit["fout"].tekst
        throw Uitleesfout(bericht: reden.isEmpty ? "HTTP \(http.statusCode)" : reden,
                          vanServer: !reden.isEmpty)
    }
    return uit.isNiets ? .boek(["type": .tekst("niets")]) : uit
}

// Wat er terugkomt is van buiten. Het wordt hier meteen hetzelfde ding als de
// rest van de app kent: wekelijks of niet, taak of niet.
private let SOORTEN = ["bijzonderheid", "stap", "weekritme"]

func schoonVoorstel(_ ruw: Json, _ mensen: [Persoon]) -> Voorstel? {
    let gevraagd = ruw["soort"].tekst
    let soort = SOORTEN.contains(gevraagd) ? gevraagd : "bijzonderheid"
    let bekend = Set(mensen.map { $0.id })
    let datum = isDatum(ruw["datum"].tekst) ? ruw["datum"].tekst : ""

    var ding = Ding()
    ding.icoon = ruw["icoon"].tekst(soort == "stap" ? "⭐" : "🎉")
    ding.tekst = ruw["tekst"].tekst.isEmpty ? ruw["label"].tekst : ruw["tekst"].tekst
    ding.wekelijks = soort == "weekritme"
    ding.taak = soort == "stap"
    ding.tijd = ruw["tijd"].tekst
    ding.tot = ruw["tot"].tekst
    ding.datum = datum
    ding.dagen = dagenVan(ruw["dagen"])
    ding.wie = ruw["wie"].lijst.map { $0.tekst }.filter { bekend.contains($0) }
    ding.ritme = ruw["ritme"].tekst == "nacht" ? .nacht : .dag
    ding.groep = ruw["groep"].tekst

    if ding.tekst.isEmpty { return nil }
    // Wat elke week terugkomt heeft dagen in plaats van een datum; leeg betekent
    // daar elke dag, dus dat mag.
    guard ding.wekelijks || !ding.datum.isEmpty else { return nil }
    return Voorstel(ding: ding, bron: String(ruw["bron"].tekst.prefix(200)))
}

// Hetzelfde bericht twee keer plakken hoort niets dubbels op te leveren.
private func dingSleutel(_ tekst: String, _ datum: String, _ dagen: [String], _ wie: [String]) -> String {
    [tekst.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
     datum.trimmingCharacters(in: .whitespacesAndNewlines),
     dagenVan(dagen).joined(separator: "+"),
     wie.joined(separator: "+")].joined(separator: "|")
}

func alBekend(_ bron: Ruw, _ g: Ding) -> Bool {
    let sleutel = dingSleutel(g.tekst, g.datum, g.dagen, g.wie)
    if g.taak {
        return bron[g.ritme].contains { groep in
            groep.stappen.contains {
                dingSleutel($0.label, $0.datum, $0.dagen, $0.wie) == sleutel
            }
        }
    }
    if g.wekelijks {
        return bron.overzicht.contains {
            dingSleutel($0.tekst, "", $0.dagen, $0.wie) == sleutel
        }
    }
    return bron.events.contains {
        dingSleutel($0.tekst, $0.datum, [], $0.wie) == sleutel
    }
}

File: mobiel/onderdelen/assistent.ts
// School, voetbal, judo en de bso sturen berichten; die wil je niet overtikken.
// Plak er een in en er komt een lijstje uit dat je met een tik overneemt. Weet
// de uitlezer iets niet — welk kind in 1-2B zit bijvoorbeeld — dan vraagt hij
// ernaar, en het antwoord blijft als kenmerk bij het kind staan.
import { lijstVan, tekst } from './inhoud';
import { dagenVan, isDatum } from './schoon';
import type { Ding } from './ding';
import type { Persoon, Ritme } from './soorten';
import type { Ruw } from './schoon';

// Op web staat de uitlezer op dezelfde Worker als de app zelf. Een telefoon
// heeft een heel adres nodig; dat komt dan uit de omgeving.
const ADRES = process.env.EXPO_PUBLIC_ASSISTENT_URL || '/api/lees';
const SLEUTEL = process.env.EXPO_PUBLIC_ASSISTENT_SLEUTEL || '';

export type Lading = {
  tekst: string; vandaag: string; ronde: number;
  kinderen: { id: string; naam: string; kenmerken: Record<string, string> }[];
};

export type Vraag = { sleutel: string; vraag: string; opties: string[]; meerkeuze: boolean };
export type Voorstel = Ding & { bron: string };

export class Uitleesfout extends Error {
  vanServer: boolean;
  constructor(bericht: string, vanServer: boolean) {
    super(bericht);
    this.vanServer = vanServer;
  }
}

export async function vraagAssistent(lading: Lading): Promise<any> {
  const kopjes: Record<string, string> = { 'Content-Type': 'application/json' };
  if (SLEUTEL) kopjes['X-Routine-Sleutel'] = SLEUTEL;
  const res = await fetch(ADRES, { method: 'POST', headers: kopjes, body: JSON.stringify(lading) });
  const uit = await res.json().catch(() => null);
  if (!res.ok) {
    // Zegt de uitlezer zelf wat er mis is, dan is dat de melding — er hoeft
    // geen 'het lukte niet' omheen.
    throw new Uitleesfout(tekst(uit && uit.fout, 'HTTP ' + res.status), Boolean(uit && uit.fout));
  }
  return uit || { type: 'niets' };
}

// Wat er terugkomt is van buiten. Het wordt hier meteen hetzelfde ding als de
// rest van de app kent: wekelijks of niet, taak of niet.
const SOORTEN = ['bijzonderheid', 'stap', 'weekritme'];

export function schoonVoorstel(ruw: any, mensen: Persoon[]): Voorstel | null {
  const soort = SOORTEN.includes(tekst(ruw?.soort)) ? tekst(ruw.soort) : 'bijzonderheid';
  const bekend = new Set(mensen.map((p) => p.id));
  const datum = isDatum(ruw?.datum) ? tekst(ruw.datum) : '';
  const uit: Voorstel = {
    icoon: tekst(ruw?.icoon, soort === 'stap' ? '⭐' : '🎉'),
    tekst: tekst(ruw?.tekst || ruw?.label),
    wekelijks: soort === 'weekritme',
    taak: soort === 'stap',
    tijd: tekst(ruw?.tijd),
    tot: tekst(ruw?.tot),
    datum,
    dagen: dagenVan(ruw?.dagen),
    wie: lijstVan<any>(ruw?.wie).map(String).filter((id) => bekend.has(id)),
    bron: tekst(ruw?.bron).slice(0, 200),
    ritme: (tekst(ruw?.ritme) === 'nacht' ? 'nacht' : 'dag') as Ritme,
    groep: tekst(ruw?.groep),
  };
  if (!uit.tekst) return null;
  // Wat elke week terugkomt heeft dagen in plaats van een datum; leeg betekent
  // daar elke dag, dus dat mag.
  return uit.wekelijks || uit.datum ? uit : null;
}

// Hetzelfde bericht twee keer plakken hoort niets dubbels op te leveren.
function dingSleutel(item: any, taak: boolean): string {
  return [tekst(taak ? item.label : item.tekst).toLowerCase(),
          tekst(item.datum), dagenVan(item.dagen).join('+'),
          lijstVan<string>(item.wie).join('+')].join('|');
}

export function alBekend(bron: Ruw, g: Ding): boolean {
  const sleutel = dingSleutel(g, g.taak);
  if (g.taak) {
    return lijstVan<any>(bron[g.ritme]).some((groep) =>
      lijstVan<any>(groep.stappen).some((stap) => dingSleutel(stap, true) === sleutel));
  }
  const lijst = g.wekelijks ? bron.overzicht : bron.events;
  return lijstVan<any>(lijst).some((item) => dingSleutel(item, false) === sleutel);
}
